src/Compilers/Core/Portable/PEWriter/MetadataHeapsBuilder.cs

```diff
@@ -269,6 +269,14 @@ public int GetUserStringToken(string str)
                 Debug.Assert(!_streamsAreComplete);
 
                 index = _userStringWriter.Position + _userStringHeapStartOffset;
+
+                // User strings are referenced by metadata tokens (8 bits of which are used for the token type) leaving only 24 bits for the offset. 
+                if ((index & 0xFF000000) != 0)
```

Comment:
This seems to be so rare that I'd consider throwing PeWritingException instead of trying to recover.